nightly.swift

```diff
@@ -2,13 +2,166 @@
+        case .dumpTimedOut:
+            return "Dump Timed Out"
+        case .networkingError(let error), .decoderError(let error), .fileSystemError(let error), .unknownError(let error):
+            return error.localizedDescription
+        case .unknownGitHost(let host):
+            return "Unknown URL host: \(host ?? "nil")"
+        case .badPackageDump(let output):
+            return "Bad Package Dump -- \(output ?? "No Output")"
+        case .missingProducts:
+            return "Missing Products"
+        case .rateLimitExceeded(let limit):
+            return "Rate Limit of \(limit) Exceeded"
+        case .packageDoesNotExist(let url):
+            return "Package Does Not Exist: \(url)"
+        case .repoIsFork:
+            // A decision has been made that the index as a whole should support forks, but not as part of dependency analysis.
+            //
+            // This is because there's an unhealthy amount of forks with a single patch to simply make the dependency work with their library,
+            // thse are often unmaintained and don't delivery huge amounts of value.
+            return "Forks are not added as part of dependency analysis"
+        case .outdatedToolsVersion:
+            return "Repo is using an outdated package format"
+        }
+    }
+}
+
+// MARK: - Networking
+
+func downloadSync(url: String, timeout: Int = 10) -> Result<Data, ValidatorError> {
+    let semaphore = DispatchSemaphore(value: 0)
+    
+    guard let apiURL = URL(string: url) else {
+        return .failure(.invalidURL(url))
+    }
+    
+    var payload: Data?
+    var taskError: ValidatorError?
+    
+    var request = URLRequest(url: apiURL)
+    
+    if let token = patToken, apiURL.host?.contains(SourceHost.GitHub.rawValue) == true {
+        request.addValue("Basic \(token)", forHTTPHeaderField: "Authorization")
+    }
+    
+    let task = session.dataTask(with: request) { (data, response, error) in
+        
+        let httpResponse = response as? HTTPURLResponse
+        
+        if let limit = httpResponse?.value(forHTTPHeaderField: "X-RateLimit-Limit").flatMap(Int.init),
+           let remaining = httpResponse?.value(forHTTPHeaderField: "X-RateLimit-Remaining").flatMap(Int.init),
+           remaining == 0 {
+            taskError = .rateLimitExceeded(limit)
+        } else if httpResponse?.statusCode == 404 {
+            taskError = .packageDoesNotExist(apiURL.absoluteString)
+        } else if let error = error {
+            taskError = .networkingError(error)
+        }
+        
+        if let dataUnwrapped = data, httpResponse?.statusCode != 200 {
+            print(String(data: dataUnwrapped, encoding: .utf8) ?? "No Data")
+        }
+        
+        payload = data
+        semaphore.signal()
+        
+    }
+    
+    task.resume()
+    
+    switch semaphore.wait(timeout: .now() + .seconds(timeout)) {
+    case .timedOut:
+        return .failure(.timedOut)
+    case .success where payload == nil:
```

Comment:
Feel like this logic is broken on the validate script too (I copied it directly from there).

GitHub will return us data even if an error is found. Payload won't be nil and therefore we won't ever attempt to return `taskError`. Let's fix that.